finetuner/tuner/callback/model_checkpoint.py

```diff
@@ -0,0 +1,141 @@
+        maximized or minimized.
+    - The frequency it should save at. Currently, the callback supports saving at
+        the end of every epoch, or after a fixed number of training batches.
+    """
+
+    def __init__(
+        self,
+        filepath: str = None,
+        save_best_only: Optional[bool] = False,
+        monitor: Optional[str] = 'val_loss',
+        mode: Optional[str] = 'auto',
+    ):
+        """
+        :param filepath: string or `PathLike`, path to save the model file.
+        :param save_best_only: if `save_best_only=True` only the best model
+        model will be saved according to the quantity monitored
+        :param monitor: if `monitor='loss'` best bodel saved will be according
+        to the training loss, if `monitor='val_loss'` best model saved will be
+        according to the validation loss
+        :param mode: one of {'auto', 'min', 'max'}. If `save_best_only=True`, the
+        decision to overwrite the current save file is made based on either
+        the maximization or the minimization of the monitored quantity.
+        For `val_acc`, this should be `max`, for `val_loss` this should be
+        `min`, etc. In `auto` mode, the mode is set to `max` if the quantities
+        monitored are 'acc' or start with 'fmeasure' and are set to `min` for
+        the rest of the quantities.
+        """
+        self.__logger = JinaLogger(self.__class__.__name__)
+        self.__filepath = filepath
+        self.__save_best_only = save_best_only
+        self.__monitor = monitor
+        if not filepath:
+            raise ValueError(
+                '``filepath`` parameter is mandatory. Pass it in parameters'
+            )
+
+        if mode not in ['auto', 'min', 'max']:
+            self.__logger.warning(
+                'ModelCheckpoint mode %s is unknown, ' 'fallback to auto mode.', mode
+            )
+            mode = 'auto'
+
+        if mode == 'min':
+            self.__monitor_op = np.less
+            self.__best = np.Inf
+        elif mode == 'max':
+            self.__monitor_op = np.greater
+            self.__best = -np.Inf
+        else:
+            if 'acc' in self.__monitor or self.__monitor.startswith('fmeasure'):
+                self.__monitor_op = np.greater
+                self.__best = -np.Inf
+            else:
+                self.__monitor_op = np.less
+                self.__best = np.Inf
+
+    def get_monitor_op(self):
+        return self.__monitor_op
+
+    def get_best(self):
+        return self.__best
+
+    def on_train_epoch_end(self, tuner: 'BaseTuner'):
+        """
+        Called at the end of the training epoch.
+        """
+        if self.__monitor == 'loss':
+            self._save_model(tuner)
+
+    def on_val_end(self, tuner: 'BaseTuner'):
+        """
+        Called at the end of the validation epoch.
+        """
+        if self.__monitor == 'val_loss':
+            self._save_model(tuner)
```

Comment:
I suggest we do 2 changes:

1. save only happens at the end of `epoch end`.
2. if we have `val_loss`, it's only make sense to save model which has a better validation loss, not training loss. otherwise, save best model with loss (incase user didn't specify val_data).
3. maybe we need to checkpoint more things, see how pytorch doing (here)[https://github.com/pytorch/examples/blob/537f6971872b839b36983ff40dafe688276fe6c3/imagenet/main.py#L237]

Reply:
I think we need both epoch end and val end because at everytime we send current loss within the tuner state, if it's epoch end I think only training loss is sent and if its val end it's the validation loss that's sent

Reply:
I think it should be on epoch end. The value in `current_loss` is simply the last loss computed - if there is validation, it will be loss of last validation loss. But as I explain in another comment, you actually need to collect validation losses for all batches and average them

Reply:
Okay I'll make changes accordingly